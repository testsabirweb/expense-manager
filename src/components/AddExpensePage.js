const React = require('react');
const ExpenseForm = require('./ExpenseForm')
const { connect } = require('react-redux')
const { addExpense } = require('./../actions/expenses')

const AddExpensePage = (props) => {
    return (
        <div>
            <h1>add expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.dispatch(addExpense(expense))
                    props.history.push('/')//for redirection to home page
                }}
            />
        </div>
    )
}

const ConnectedAddExpensePage = connect()(AddExpensePage)
module.exports = ConnectedAddExpensePage