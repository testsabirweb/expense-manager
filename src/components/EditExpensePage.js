const React = require('react');
const { connect } = require('react-redux')
const ExpenseForm = require('./ExpenseForm')
const { editExpense, startRemoveExpense } = require('./../actions/expenses')

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')//for redirectonto home page
    }
    onClick = (e) => {
        this.props.startRemoveExpense({ id: this.props.expense.id })
        this.props.history.push('/')//for redirectonto home page
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onClick}
                > REMOVE
                </button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}
const mapDispatchToProps = (dispatch,props) => {
    return {
        editExpense:(id,expense)=>dispatch(editExpense(id,expense)),
        startRemoveExpense:(data)=>dispatch(startRemoveExpense(data))
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)