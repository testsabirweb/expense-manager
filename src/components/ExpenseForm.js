const React = require('react')
const moment = require('moment')
const { SingleDatePicker } = require('react-dates')

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => {
            return {
                description
            }
        })
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => {
            return {
                note
            }
        })
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => {
                return {
                    amount
                }
            })
        }
    }
    //https://github.com/airbnb/react-dates
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => {
                return {
                    createdAt
                }
            });
        }
    };
    //https://github.com/airbnb/react-dates
    onFocusChange = ({ focused }) => {
        this.setState(() => {
            return {
                calendarFocused: focused
            }
        });
    };
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState(() => {
                return {
                    error: 'please provide description and amount'
                }
            })
        }
        else {
            this.setState(() => {
                return {
                    error: ''
                }
            })
            this.props.onSubmit({///this props can come from AddExpensePage or EditExpensePage
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),//valueOf()is a method from moment.js which returns the unix time stamp
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="add note to expense(optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button>add expense</button>
                </form>
            </div>
        )
    }
}

module.exports = ExpenseForm