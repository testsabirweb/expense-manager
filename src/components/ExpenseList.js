const React = require('react')
const { connect } = require('react-redux')
const ExpenseListItem = require('./ExpenseListItem')
const selectExpenses = require('./../selectors/expenses')


export const ExpenseList = (props) => {
    return (
        <div>
            {
                props.expenses.length === 0 ? (
                    <p>No expenses</p>) : (
                        props.expenses.map((expense) => {
                            return <ExpenseListItem key={expense.id} {...expense} />
                        }))
            }
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList)

// const ConnectedExpenseList = connect((state) => {
//     return {
//         name: 'sabir'
//     }
// })(ExpenseList)

// module.exports = ConnectedExpenseList