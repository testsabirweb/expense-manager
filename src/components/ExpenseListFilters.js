const React = require('react')
const { connect } = require('react-redux')
const {
    setTextFilter,
    sortByAmount,
    sortByDate,
    setEndDate,
    setStartDate
} = require('./../actions/filters')
const { DateRangePicker } = require('react-dates')

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => {
            return {
                calendarFocused
            }
        })
    }
    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}/////here we set the value of an element
                    onChange={(e) => {
                        this.props.dispatch(setTextFilter(e.target.value));/////here value is accessed and then dispatched
                    }}
                />

                <select
                    value={this.props.filters.sortBy}
                    onChange={(e) => {
                        if (e.target.value === "date") {
                            this.props.dispatch(sortByDate())
                        }
                        else if (e.target.value === "amount") {
                            this.props.dispatch(sortByAmount())
                        }
                    }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        )

    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const ConnectedExpenseListFilters = connect(mapStateToProps)(ExpenseListFilters)

module.exports = ConnectedExpenseListFilters