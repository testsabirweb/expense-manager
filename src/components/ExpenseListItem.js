const React = require('react')
import moment from 'moment'
const { Link } = require('react-router-dom')
import numeral from 'numeral'

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div>
            <Link to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>
            â‚¹ {numeral(amount).format('0,0.00')}
            -
            {moment(createdAt).format('DD/MMM/YYYY')}
            </p>
        </div>
    )
}



module.exports = ExpenseListItem