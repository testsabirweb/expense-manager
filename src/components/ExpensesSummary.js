import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'
import selectExpenses from './../selectors/expenses'
import selectExpensesTotal from './../selectors/expense-total'
export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses'
    const formatedExpensesTotal = numeral(expensesTotal).format('0,0.00')
    return (
        <div className="page-header">
            <div className="content-container">
                <h2 className="page-header__title"> viewing <span>{expensesCount}</span> {expenseWord} totalling of â‚¹ <span>{formatedExpensesTotal}</span></h2>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)

    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)