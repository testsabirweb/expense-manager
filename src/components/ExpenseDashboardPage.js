const React = require('react');
const ExpenseList = require('./ExpenseList')
const ExpenseListFilters = require('./ExpenseListFilters');
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => {
    return (
        <div>
            <h1>
                <ExpensesSummary />
                <ExpenseListFilters />
                <ExpenseList />
            </h1>
        </div>
    )
}

module.exports = ExpenseDashboardPage
