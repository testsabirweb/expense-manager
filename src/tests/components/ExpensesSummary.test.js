import React from 'react'
import {shallow} from 'enzyme'
import{ ExpensesSummary }from './../../components/ExpensesSummary'

test('should render expense summary with 1 expense ', () => {
    const wrapper=shallow(<ExpensesSummary expensesCount={1} expensesTotal={145} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render expense summary with multiple expense ', () => {
    const wrapper=shallow(<ExpensesSummary expensesCount={10} expensesTotal={147545556} />)
    expect(wrapper).toMatchSnapshot()
})