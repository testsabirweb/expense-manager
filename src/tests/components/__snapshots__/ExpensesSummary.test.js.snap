// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with 1 expense  1`] = `
<div>
  viewing 
  1
   
  expense
   totalling 
  $145.00
</div>
`;

exports[`should render expense summary with multiple expense  1`] = `
<div>
  viewing 
  10
   
  expenses
   totalling 
  $147,545,556.00
</div>
`;
